### ISSUE TITLE
breaking changes based on CRAN v1.30.0

### ISSUE THREAD

**Issue by evanbiederstedt** [igraph/rigraph#517]
Hi there

**Describe the bug**
I've noticed breaking changes with respect to the latest version of igraph uploaded to CRAN today (1 April 2022)

https://github.com/kharchenkolab/leidenAlg/issues/9

Our R packages have been stable for a few years....it seems like there have been some changes to igraph 1.2.11 to 1.30.0 causing issues. 

Could you summarize what major changes took place? 

The next question would be for me to figure out how to work with this; I'd need to figure out how much of our current code needs to be changed, as we internally link to igraph often. 

**Version information**
v1.30.0

Any help here appreciated.

**Comment by ntamas** [comment]
It seems like `leidenAlg` is messing around with igraph's internal RNG and tries to set it to `igraph_rng_mt19937` forcefully. This is not good because the `igraph` module in R is set up in a way that it uses R's built-in RNG to integrate nicely with R. But this is not the cause of the issue; the cause is probably that from igraph 1.3.0, we control the visiblity of entry points in the compiled shared library [according to CRAN recommendations](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Controlling-visibility) and `igraph_rng_mt19937` is probably no longer accessible from the outside.

For what it's worth, revdepchecks were run on all the >900 packages with igraph 1.3.0 and this issue was not revealed on my local machine so there could be something else going on as well. I was testing on macOS and maybe the visibility rules are not applied there during the build process.

**Comment by ntamas** [comment]
Also, @vtraag may provide more information here as it seems like `leidenAlg` is a wrapper around his `leidenalg` implementation -- which is now part of igraph itself.

**Comment by szhorvat** [comment]
> the cause is probably that from igraph 1.3.0, we control the visiblity of entry points in the compiled shared library [according to CRAN recommendations](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Controlling-visibility) and `igraph_rng_mt19937` is probably no longer accessible from the outside.

This certainly seems to be the cause.

However, accessing the igraph C functions this way, by trying to load them from the shared library of R/igraph, is a fragile hack that can't be relied on. The only robust solution would be to use a separate, private copy of the igraph C library, and not try to depend on the internals of the shard library installed with R/igraph.

Note that R/igraph now does include `cluster_leiden()` but I'm not sure if it provides all the feaures that leidenAlg does (@vtraag ?)

**Comment by vtraag** [comment]
Yes, the Python `leidenalg` package vendors the `igraph` library, similar to what `python-igraph` does. I try to ensure that the Python `leidenalg` package is compatible with whatever version of `python-igraph` is released and properly indicate this in the requirements. In this case, the R version of the Python `leidenalg` package, called `leidenAlg` probably needs to adapt this in order to properly work with `igraph` in R, and simply use R's own RNG.

Perhaps we should coordinate this a little bit in the near future in order to simplify things as much as possible, I'm not directly sure what would be the best approach here.

**Comment by evanbiederstedt** [comment]
Thanks for the help everyone. I'm happy to try to include a local version of `cigraph` in `/src`, compile this, and then link to this. That appears to be the safest approach. (I don't think removing `igraph_rng_mt19937` alone is the source of the problem; other data structures from igraph we were using are also no longer accessible.)

I'm getting a bit lost in the branches between the repos though (which differ quite a bit). If it's possible to request a bit of advice/help here:

(1) Where is a stable version of cigraph I could include in `/src`? A version whereby there are no restrictions on the entry points?
(2)  Normally if there are two independent self-contained libraries within `/src` (e.g. `/src/cigraph`, `/src/leidenalg`), I would use Makefiles within each of these self-contained libraries to separately compile them, and then a Makevars file in `/src` as expected. CRAN and autotools/cmake get a bit hairy. If you do have some ideas how to easily compile cigraph within the constraints of R, please let me know---perhaps there's a simple approach here.

As an academic question:
> the cause is probably that from igraph 1.3.0, we control the visiblity of entry points in the compiled shared library [according to CRAN recommendations](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Controlling-visibility) and igraph_rng_mt19937 is probably no longer accessible from the outside.

Did CRAN specifically complain about this? This issue came out of nowhere for me, and now I'm scrambling to fix this so the package isn't removed :)

Thanks, Evan

**Comment by ntamas** [comment]
The restriction stems from `src/Makevars.in` in the source code of R-igraph, especially these lines:

```
PKG_FLAGS=$(C_VISIBILITY)
PKG_CXXFLAGS=$(CXX_VISIBILITY)
PKG_FFLAGS=$(F_VISIBILITY)
```

These lines add the compiler flags used by R when compiling igraph to ensure that everything is kept internal by default unless we export it explicitly. If you simply vendor igraph's source code in your own project and compile it along with your own code, there is nothing preventing you from accessing igraph's internals, so you don't need any special unrestricted version of the code, just grab the version you need from https://github.com/igraph/igraph . Note that igraph now uses CMake as the build system so you cannot just embed the source code as is; I'd probably advise you to take a look at the `master` branch of https://github.com/igraph/rigraph and look in the `src` folder to see how igraph is embedded in R-igraph without using CMake.

In this sense, your package is not that different from R-igraph itself; R-igraph embeds the C core of igraph and builds a bunch of R functions on top of it, while yours will also embed the C core of igraph and build a bunch of other R functions on top of it. The way we do it in https://github.com/igraph/rigraph should work the same. (Don't look at the `dev` branch of the R-igraph repo, though; we generate the code of most of the R interface automatically, and the `master` branch is simply generated from the `dev` branch, but you don't need that machinery as you are not building interfaces for all the functions in the C igraph core).

> Did CRAN specifically complain about this?

No, but it's standard practice, especially in environments where a large REPL app like R is importing third-party modules compiled into DLLs or shared libraries as you want to avoid having two functions with the same name from two different libraries.

If this is causing significant headache to you, we can restore the visibility of all functions as a temporary measure for the 1.3.x branch if CRAN agrees. Please keep us in the loop with any correspondence that you have with CRAN so we can chime in. But, I think eventually CRAN people will also insist on using the visibility flags when possible so we probably cannot postpone it too much even if we decide to revert it now.

**Comment by szhorvat** [comment]
> If this is causing significant headache to you, we can restore the visibility of all functions as a temporary measure for the 1.3.x branch if CRAN agrees.

@evanbiederstedt Just to be clear: even if we do this, you are accessing igraph internals (non-public API) in fragile ways. If you do this, leidenAlg will very likely crash the R session whenever an incompatible version of igraph is installed. It is not a long-term solution.

**Comment by szhorvat** [comment]
> But, I think eventually CRAN people will also insist on using the visibility flags when possible so we probably cannot postpone it too much even if we decide to revert it now.

To be fair, R's build system doesn't even use the visibility flags on macOS, even though macOS has supported them for a long time now ...

**Comment by evanbiederstedt** [comment]
Thanks for the help @ntamas @szhorvat 

> I'd probably advise you to take a look at the master branch of https://github.com/igraph/rigraph and look in the src folder to see how igraph is embedded in R-igraph without using CMake.

Ok, excellent. This is what I needed to know; I'll use https://github.com/igraph/rigraph/tree/master

Talking this out aloud: https://github.com/kharchenkolab/leidenAlg/tree/main/src

I'd structure this as:
```
leidenAlg/src
leidenAlg/src/leidenalg
leidenAlg/src/rigraph 
``` 
whereby `leidenAlg/src/rigraph` contains the contents of https://github.com/igraph/rigraph/tree/master/src. 

I guess I now need to write an appropriate Makefile and Makefile.win for this code, and compile from there. It looks like there are already generated `configure` files there; maybe it's best to remove them for my purposes...

Like many R packages, it's naturally less robust than a solution with cmake, but we tell users in the REAMDE + DESCRIPTION to install the required dependencies, e.g. libxml2

Disagreements? Obvious errors in my understanding above?

> The proper long-term solution is for leidenAlg to link statically to its separate, private copy of C igraph.

Agreed. I'll pursue this approach above. 

> No, but it's standard practice, especially in environments where a large REPL app like R is importing third-party modules compiled into DLLs or shared libraries as you want to avoid having two functions with the same name from two different libraries.

Fair enough, this makes sense

> If this is causing significant headache to you, we can restore the visibility of all functions as a temporary measure for the 1.3.x branch if CRAN agrees. Please keep us in the loop with any correspondence that you have with CRAN so we can chime in. But, I think eventually CRAN people will also insist on using the visibility flags when possible so we probably cannot postpone it too much even if we decide to revert it now.

As a short-term solution, it wouldn't hurt to be honest. The way this transpired is I woke up to an email from CRAN that they were pulling all of my packages (leidenAlg and packages which depend on leidenAlg) along with GitHub issues. It was a crisis that came out of nowhere :)

That said, I think if I get the above working ASAP and re-version to CRAN, all will be calm again.

**Comment by ntamas** [comment]
> `leidenAlg/src/rigraph` contains the contents of https://github.com/igraph/rigraph/tree/master/src

You don't need all the files from there; you probably only need `src/core`, which is the C core of igraph, and `src/include`, which contains the igraph headers. `init.c`, `lazyeval.c`, `rinterface.c` and `rinterface_extra.c` contains additional glue code between C-igraph and R, but these are relevant only if you want to expose the entire igraph library to R.

One additional piece of code that might be relevant is `src/rrandom.c`, which contains an igraph RNG implementation that uses R's built-in RNG to generate random numbers. This is not strictly required, but it is needed if you want `set.seed()` to have any effect on igraph's internal RNG.

> I guess I now need to write an appropriate Makefile and Makefile.win for this code, and compile from there. It looks like there are already generated configure files on the master branch; maybe it's best to remove them for my purposes...

Yes, do not look at those; igraph's `configure` file is hand-rolled because of several optional dependencies, but as you only need igraph to get `leidenAlg` working, you don't need those and you can probably just use the standard build mechanism that R provides.

> As a short-term solution, it wouldn't hurt to be honest. The way this transpired is I woke up to an email from CRAN that they were pulling all of my packages (leidenAlg and packages which depend on leidenAlg) along with GitHub issues.

Oh yeah, we also get those all the time :) Last time we were given two weeks to fix igraph because some completely harmless compiler warnings started to appear with R-devel and gcc 12 on Debian.

Can you send an email to CRAN maintainers explaining the situation and adding me in cc? (You can find my email address on igraph's CRAN page). I can revert the visibility changes in 1.3.1 but only if CRAN maintainers are okay with that.

**Comment by evanbiederstedt** [comment]
Thanks for the continued help here @ntamas 

It looks like this branch now works, `/feature/compile_rigraph`: https://github.com/kharchenkolab/leidenAlg

There is one file which didn't compile correctly, so I excluded it. `core/centrality/prpack.c`. I always get:

```
core/centrality/prpack.cpp:95:9: error: unknown type name 'prpack_igraph_graph'; did you mean 'prpack_base_graph'?
```

I also needed to include `/vendor` for things like `cs/cs.h` and `plfit`. 

Looks like it works on Mac OS and Windows as well. 

I'll doublecheck tomorrow if I'm missing something, but maybe this is fixed.

**Comment by szhorvat** [comment]
`PRPACK_IGRAPH_SUPPORT` needs to be defined when compiling.

You're not in favour of the simple solution of using `cluster_leiden` then? https://github.com/kharchenkolab/leidenAlg/issues/9#issuecomment-1091845258

**Comment by szhorvat** [comment]
> `init.c`, `lazyeval.c`, `rinterface.c` and `rinterface_extra.c` contains additional glue code between C-igraph and R, but these are relevant only if you want to expose the entire igraph library to R.

@ntamas Isn't there a risk of conflict with igraph itself if all the glue code is included, in particular the initializer?

https://github.com/kharchenkolab/leidenAlg/blob/93931ce087c04bf4b325596192cf94e757e49043/src/rigraph/rinterface.c#L2461

I assume the R/igraph initializer will not be run from leidenAlg (which would be bad)? How does R decide which function to run? Does it just glue together `R_init_` with the package name?

----

There's also the issue of setting up proper error handlers for this copy of igraph, so that passing in wrong input won't just abort. In fact, all the `#ifdef USING_R` means that the abort code is disabled, and _by default_ there is no proper error handling (e.g. there is no freeing of memory). When `USING_R` is defined during compilation, the assumption is that an alternative error handler will be set up manually.

**Comment by evanbiederstedt** [comment]
> You're not in favour of the simple solution of using cluster_leiden then?

Well....let me try to parse our interaction for a moment. I need to reflect on the past 7 days.

Background: I wake up to GitHub issues and emails about the CRAN packages breaking. So I need to drop what I'm doing to fix this. 

I read this:

> However, accessing the igraph C functions this way, by trying to load them from the shared library of R/igraph, is a fragile hack that can't be relied on, and was never guaranteed to work. The only robust solution would be to use a separate, private copy of the igraph C library, and not try to depend on the internals of the shard library installed with R/igraph.

> @evanbiederstedt Just to be clear: even if we do this, you are accessing igraph internals (non-public API) in fragile ways. If you do this, leidenAlg will very likely crash the R session whenever an incompatible version of igraph is installed. It is not a long-term solution. The proper long-term solution is for leidenAlg to link statically to its separate, private copy of C igraph. (Things will get better once https://github.com/igraph/rigraph/issues/268, but there's a long way until that becomes possible.)

Ok, so for the past two years, we used an unstable hack. We made a mistake. We need to fix this immediately.

The solution above recommended by you would be to pursue a separate, private copy and compile this. So I've done this. 

There was another idea you mentioned 18 hours ago: https://github.com/kharchenkolab/leidenAlg/issues/9#issuecomment-1091705355

Other ideas: unfortunately, we cannot delete the package entirely, due to CRAN licensing. "Archiving" for CRAN doesn't work either, as the mirrors and URLs still exist. 

One question I had to myself is....if we release a package which is just a wrapper around a single function of your active code base which is being worked on daily, is this not also an unstable hack? 

I'm not really sure what the correct answer is. There are several unofficial R packages trying to implement an R version of the Python code: https://github.com/vtraag/leidenalg#usage

This is largely driven by people in computational biology. Our package also doesn't implement the full functionality for the python version of `leidenalg`. (At one point @vtraag and I were discussing implementing the full functionality of the python package in R but...there's other work to do---apologies again Vincent.)

So...yeah, what is the right move here? What's the best investment in time?

**Comment by szhorvat** [comment]
> One question I had to myself is....if we release a package which is just a wrapper around a single function of your active code base which is being worked on daily, is this not also an unstable hack?

No, using documented, public API is not a hack.

**Comment by szhorvat** [comment]
> There was another idea you mentioned 18 hours ago: [kharchenkolab/leidenAlg#9 (comment)](https://github.com/kharchenkolab/leidenAlg/issues/9#issuecomment-1091705355)

I was looking for ways to save you some trouble. At the point where you are effectively setting up your own build system for igraph (instead of using CMake), things will get complicated and error-prone. CRAN seems to make it a bit difficult to just use CMake, and even if they didn't, they might complain about the `abort` and `printf` symbols present in igraph, even if these functions are never actually called. So you probably _will_ be forced to use a custom approach to building, as you are doing here, and as `rigraph` is also doing. Doing that in a robust way is not easy.

I can definitely see the advantage of using Vincent's leidenalg code: in the future you can choose to expose more of its functionality. It is up to you to decide if this future potential is worth all the extra work that you'll need to put in now to make everything work robustly. As I said in my earlier comment, it seems to me that it's not quite working correctly at the moment.

**Comment by szhorvat** [comment]
> There's also the issue of setting up proper error handlers for this copy of igraph

I looked more carefully at the leidenalg code, and it seems that it does not use error handling (it is not checking error codes), so this should not be a problem. Presumably then the only way it can fail is an out-of-memory condition, which is unlikely. In other words, presumably it does not rely on igraph functions checking their input and throwing an error if it is invalid, and instead it ensures not to pass in invalid input.

So in summary, it's not necessary to set up an error handler after all @evanbiederstedt

**Comment by szhorvat** [comment]
I keep looking at where else there might be problems. Something that caught my eye is this line:

https://github.com/kharchenkolab/leidenAlg/blob/71bbf163d216bcd0321e4a585a4d864baa8fac41/src/leiden.cpp#L38

This creates an `igraph_t` object for your vendored igraph library, and fills its attributes from R/igraph. However, the igraph used by leidenAlg is not set up to use R/igraph attributes. If attributes are accessed, it is likely that there will be a crash.

Similar to the error handling case, I _expect_ (but have not checked!) that the leidenalg code does not access attributes at all (unless this is done in some implicit way?) Thus the problem _should_ not be triggered.

But in the end the `attr` member should not be filled in if there is no attribute handler—just keep it as `NULL`.

**Comment by ntamas** [comment]
> Isn't there a risk of conflict with igraph itself if all the glue code is included, in particular the initializer?

`rinterface.c` should be left out for sure; it's not part of the core C igraph, it's part of the glue code between R and C-igraph and it is responsible for registering functions in the `igraph` module, so that's definitely not needed for `leidenAlg`.

**Comment by szhorvat** [comment]
Re attribute handling, there's this comment as well: https://github.com/kharchenkolab/leidenAlg/blob/main/src/leidenalg/Optimiser.cpp#L8 So in fact it does use the attribute handler and needs one set up properly.

**Comment by ntamas** [comment]
Okay, but then the problem is the conversion of attributes from R to C. If I understand correctly, `leidenAlg` takes the raw R representation of an igraph graph (which is essentially a list), and fills an `igraph_t` object with the vectors taken from the raw R representation, just like the official R interface of igraph does it. This probably works fine for the edge list and the related indexes. The problem is the attribute handling: if you attach the C attribute handler to the vendored igraph in `leidenAlg`, then oyu need to provide a data structure in `g->attr` that the C attribute handler understands. Obviously this is not possible if you directly take the pointer of the R attribute storage object and snap it to `g->attr`. The alternative would be to attach the R attribute handler to the vendored igraph, but we don't expose that either due to the visibility settings.

If you only need a weight vector in leidenAlg, I'd probably extract that weight vector first in R into a variable, and then pass that variable directly into a lower-level C function that could then attach the C attribute handler of igraph and process the weight vector using the C attribute handler. Or, even better, just pass the weight vector directly to the underlying Leiden algorithm implementation if that is possible, avoiding any use of the attribute handler interface.

**Comment by szhorvat** [comment]
> The problem is the attribute handling: if you attach the C attribute handler to the vendored igraph in `leidenAlg`, then oyu need to provide a data structure in `g->attr` that the C attribute handler understands.

Does leidenalg require existing attributes in the graph passed to it, or does it just make use of the attribute handler for internal operations? It seems to me that it does the latter. In that case all that needs to be done is to construct the new `igraph_t` using documented means (`igraph_create()`) and not modify its internals, as it currently done. This is another reason why this solution is fragile: it depends on internal details of both `igraph_t` as well as the `igraph` class in R.


> If you only need a weight vector in leidenAlg, I'd probably extract that weight vector first in R into a variable, and then pass that variable directly into a lower-level C function that could then attach the C attribute handler of igraph and process the weight vector using the C attribute handler. Or, even better, just pass the weight vector directly to the underlying Leiden algorithm implementation if that is possible, avoiding any use of the attribute handler interface.

👍

**Comment by vtraag** [comment]
> Re attribute handling, there's this comment as well: https://github.com/kharchenkolab/leidenAlg/blob/main/src/leidenalg/Optimiser.cpp#L8 So in fact it does use the attribute handler and needs one set up properly.

Sorry this threw you off, it is an old comment in the code that should be remove. Internally, my code doesn't use any attribute handler from `igraph` and works directly with vectors of weights without using any attribute handler. I considered using an attribute handler at one point, but it was actually more challenging to work with at the time (I don't remember why anymore).

> Or, even better, just pass the weight vector directly to the underlying Leiden algorithm implementation if that is possible, avoiding any use of the attribute handler interface.

Yes, exactly, this is what I do in the Python interface as well: the Python side gets the attributes from the graph and passes it simply directly as a vector to the underlying C implementation.

> This is another reason why this solution is fragile: it depends on internal details of both `igraph_t` as well as the `igraph` class in R.

This is perhaps true, but also unavoidable. If we have to handle an R `igraph` object directly in `C`, we have no way around creating an `igraph_t` object somehow from the `R` representation. In Python I also need to do the same, and call `PyCapsule_GetPointer` in order to get the underlying `igraph_t` object of the Python `Graph` object. Luckily, in Python this is fairly straightforward, and at least we can directly use `igraph_t` and just represent it as a pointer, but unfortunately in R we have to go through this different representation as vectors, since that is how `igraph` in R works. Perhaps we should make a *documented* and *officially supported* way of obtaining a C `igraph_t` object from an R `igraph` object, and vice-versa, create an R `igraph` object from a C `igraph_t` object. Without it, such hacks will continue to be necessary in one way or another.

> The proper long-term solution is for leidenAlg to link statically to its separate, private copy of C igraph. (Things will get better once [R/igraph starts to use a separate C/igraph](https://github.com/igraph/rigraph/issues/268), but there's a long way until that becomes possible.)

This will help a long way to improve the situation. There is one big advantage to this: as far as I know, for system libraries, CRAN rules don't apply. That is, all the questions about `printf` statements, `rand` functions and `abort` all don't directly apply to such system libraries, as far as I understand. Hence, it should make it easier to make `igraph` available as a pure C library with then a wrapper on R that is provided through CRAN. Given that Windows now relies on `RTools` which is essentially just a variant of `msys2`, it should not be too difficult to provide `igraph` as a system library for all platforms in this way. Of course the real challenge is then the versioning, which will become a separate nightmare, as least as long as we don't have a stable 1.0 version of the C core yet. Nonetheless, even then, we still face the problem of how to interface between R and C.

> Our package also doesn't implement the full functionality for the python version of `leidenalg`. (At one point @vtraag and I were discussing implementing the full functionality of the python package in R but...there's other work to do---apologies again Vincent.)
> 
> So...yeah, what is the right move here? What's the best investment in time?

For now, I'd say that people should just use `cluster_leiden` directly from `igraph`, especially given that the current `leidenAlg` implementation is relatively limited anyway. This suggestion by Szabolcs is true though:

> I can definitely see the advantage of using Vincent's leidenalg code: in the future you can choose to expose more of its functionality. It is up to you to decide if this future potential is worth the extra work that you'll need to put in now to make everything work robustly. As I said in my earlier comment, it seems to me that it's not quite working correctly at the moment.

But I think it will be a long shot to get this working now. Perhaps in the future something like this might be possible, but unfortunately we all seem to be stretched for time.

**Comment by evanbiederstedt** [comment]
Thanks for everyone's help here. I think we have this resolved. 

I put a note in the README here: https://github.com/kharchenkolab/leidenAlg#leidenalg

I wish I had more time to make the library fully compatible with the python bindings, but I currently don't. 

Academic aside:

> No, using documented, public API is not a hack.

Sorry if the rhetorical question came across as grumpy; I realize it's not a hack of course because there's a version associated with it. The documented, public API itself doesn't matter though---we change them all the time (but we version them and try to make them backwards compatible). My point is, depending on the future directions you'd like to take igraph + leiden clustering, that function may change, much like linking to a shared library. (This is a library that's certainly in active development, as this conversation shows + 30 branches on Github + almost daily commits.) It's a fundamental problem with any dependency within CRAN----there's some level of trust I have with any dependency that it would change too much to cause me problems (i.e. dev time to fix). 

I'll agree that using a documented function is more stable than linking to a library though, and we can leave it at that.
